src/inmanta/deploy/state.py

```diff
@@ -201,71 +166,120 @@ def add_up_to_date_resource(self, resource: ResourceIdStr, details: ResourceDeta
                 self.resource_state[resource].status = ResourceStatus.UNDEFINED
             self.resource_state[resource].blocked = BlockedStatus.YES
         else:
-            resource_status = ResourceStatus.UNDEFINED if not transient else ResourceStatus.HAS_UPDATE
+            resource_status = ResourceStatus.UNDEFINED if not is_transitive else ResourceStatus.HAS_UPDATE
             self.resource_state[resource] = ResourceState(
                 status=resource_status, deployment_result=DeploymentResult.NEW, blocked=BlockedStatus.YES
             )
             self.types_per_agent[details.id.agent_name][details.id.entity_type] += 1
         self.dirty.discard(resource)
 
-    def block_provides(self, resources: Set[ResourceIdStr]) -> set[ResourceIdStr]:
+    def block_provides(self, resources: Set[ResourceIdStr]) -> Set[ResourceIdStr]:
         """
         Marks the provides of the given resources as blocked.
 
         Must be called under the scheduler lock. This method assumes that all the resources of the model version
         are populated in the resources dictionary.
 
         :param resources: The set of resources for which the provides have to be blocked.
-        :return: The set of dependent resources that were marked as blocked (transitively).
+        :return: The set of dependent resources that were newly marked as blocked (transitively).
         """
-        result = set()
-        for dependent_resource in self.requires.get_all_provides_transitively(resources):
-            self.block_resource(dependent_resource, self.resources[dependent_resource], transient=True)
-            result.add(dependent_resource)
+        result: set[ResourceIdStr] = set()
+        provides_view: Mapping[ResourceIdStr, Set[ResourceIdStr]] = self.requires.provides_view()
+        todo: list[ResourceIdStr] = list(itertools.chain.from_iterable(provides_view.get(r, set()) for r in resources))
+        # We rely on the seen set to improve performance. Although the improvement might be rather small,
+        # because we only save one call to `self.resource_state[resource_id].blocked.is_blocked()`.
+        seen: set[ResourceIdStr] = set()
+        while todo:
+            resource_id: ResourceIdStr = todo.pop()
+            if resource_id in seen:
+                continue
+            seen.add(resource_id)
+            if self.resource_state[resource_id].blocked.is_blocked():
+                # Resource is already blocked. All provides will be blocked as well.
+                continue
+            else:
+                # Block resource and its provides.
+                self.block_resource(resource_id, self.resources[resource_id], is_transitive=True)
+                result.add(resource_id)
+                todo.extend(provides_view.get(resource_id, set()))
         return result
 
-    def unblock_resource(self, resource: ResourceIdStr) -> None:
+    def mark_as_defined(self, resource: ResourceIdStr, details: ResourceDetails) -> None:
```

Comment:
I think it might be beneficial to do this in bulk for all relevant resources so we can share the cache, but we'll probably get that for free when we fix the bug we discussed earlier, so I don't think we need to at the moment.